pkg_building chapter Portuguese translation (#952)
ropensci/dev_guide: pkg_building.pt.Rmd

---
aliases:
- building.html
---

> # Guia de embalagem {#building}

Reviewer: ```suggestion
# Guia de desenvolvimento de pacotes {#building}
```
@fblpalmeira  a tradução mais "direta" seria algo como "guia de empacotamento", mas não parece muito claro...

```{block, type="summaryblock"}
A rOpenSci aceita pacotes que atendam às nossas diretrizes por meio de um processo simplificado de [Revisão por Pares de Software] (#whatissoftwarereview). Para garantir um estilo consistente em todas as nossas ferramentas, escrevemos este capítulo destacando nossas diretrizes para o desenvolvimento de pacotes. Leia e aplique também nosso [capítulo sobre integração contínua (CI)](#ci). Outras orientações para depois do processo de revisão são fornecidas na terceira seção deste livro, começando com [um capítulo sobre colaboração](#collaboration).

Recomendamos que os desenvolvedores de pacotes leiam o livro completo de Hadley Wickham e Jenny Bryan sobre desenvolvimento de pacotes, que está disponível [gratuitamente on-line](https://r-pkgs.org/). Nosso guia é parcialmente redundante em relação a outros recursos, mas destaca as diretrizes da rOpenSci.

>  Para saber por que vale a pena enviar um pacote para a rOpenSci para atender às diretrizes, dê uma olhada em [reasons to submit](#whysubmit). 

Reviewer: ```suggestion
 Para saber por que vale a pena enviar um pacote para a rOpenSci para atender às diretrizes, dê uma olhada em [motivos para submeter](#whysubmit). 
```

```

## Nome do pacote e metadados {#package-name-and-metadata}

### Como nomear seu pacote {#naming-your-package}

> - Recomendamos fortemente nomes curtos e descritivos em letras minúsculas. Se o seu pacote tratar de um ou mais serviços comerciais, verifique se o nome não viola as diretrizes de marca. Você pode verificar se o nome do seu pacote está disponível, é informativo e não é ofensivo usando o [`pak::pkg_name_check()` função](https://pak.r-lib.org/reference/pkg_name_check.html) use também um mecanismo de pesquisa para ver se é ofensivo em um idioma diferente do inglês. Em particular, faça *não* escolha um nome de pacote que já esteja sendo usado no CRAN ou no Bioconductor.

Reviewer: ```suggestion
- Recomendamos fortemente nomes curtos e descritivos em letras minúsculas. Se o seu pacote tratar de um ou mais serviços comerciais, verifique se o nome não viola as diretrizes de marca. Você pode verificar se o nome do seu pacote está disponível, é informativo e não é ofensivo usando a função [`pak::pkg_name_check()`](https://pak.r-lib.org/reference/pkg_name_check.html); use também um mecanismo de pesquisa para ver se é ofensivo em um idioma diferente do inglês. Em particular, *não* escolha um nome de pacote que já esteja sendo usado no CRAN ou no Bioconductor.
```

- Há uma compensação entre as vantagens de um nome de pacote exclusivo e um nome de pacote menos original.
  
  - Um nome de pacote mais exclusivo pode ser mais fácil de rastrear (para você e nós avaliarmos o uso do pacote, por exemplo, menos falsos positivos ao digitar seu nome na pesquisa de código do GitHub) e pesquisar (para os usuários perguntarem "como usar o pacote blah" em um mecanismo de pesquisa).

>   - Por outro lado, um *também* único demais pode fazer com que o pacote seja menos detectável (ou seja, não é possível encontrá-lo pesquisando "como fazer isso em R"). Isso pode ser um argumento para nomear seu pacote com algo muito próximo ao tópico, como [geojson](https://github.com/ropensci/geojson)).

Reviewer: ```suggestion
  - Por outro lado, um nome único *demais* pode fazer com que o pacote seja menos detectável (ou seja, não seja possível encontrá-lo pesquisando "como fazer isso em R"). Isso pode ser um argumento para nomear seu pacote com algo muito próximo ao tópico, como [geojson](https://github.com/ropensci/geojson)).
```

> - Encontre outros aspectos interessantes sobre como nomear seu pacote [nesta postagem do blog de Nick Tierney](https://www.njtierney.com/post/2018/06/20/naming-things/) e, caso você mude de ideia, descubra [como renomear seu pacote nesta outra postagem do blog de Nick](https://www.njtierney.com/post/2017/10/27/change-pkg-name/).

Reviewer: ```suggestion
- Encontre outros aspectos interessantes sobre como nomear seu pacote [neste texto do blog do Nick Tierney (em inglês)](https://www.njtierney.com/post/2018/06/20/naming-things/) e, caso você mude de ideia, descubra [como renomear seu pacote nesta outra postagem do blog do Nick](https://www.njtierney.com/post/2017/10/27/change-pkg-name/).
```

> ### Criação de metadados para seu pacote {#creating-metadata-for-your-package}

Reviewer: ```suggestion
### Criando metadados para seu pacote {#creating-metadata-for-your-package}
```

> Recomendamos que você use o [`codemetar` pacote](https://github.com/ropensci/codemetar) para criar e atualizar um arquivo JSON [CodeMeta](https://codemeta.github.io/) para seu pacote por meio de `codemetar::write_codemeta()`. Ele incluirá automaticamente todas as informações úteis, incluindo [Tópicos do GitHub](#grooming). O CodeMeta usa [Termos do Schema.org](https://schema.org/) portanto, à medida que ganha popularidade, os metadados JSON do seu pacote podem ser usados por serviços de terceiros, talvez até por mecanismos de pesquisa.

Reviewer: ```suggestion
Recomendamos que você use o [pacote `codemetar`](https://github.com/ropensci/codemetar) para criar e atualizar um arquivo JSON [CodeMeta](https://codemeta.github.io/) para seu pacote por meio de `codemetar::write_codemeta()`. Ele incluirá automaticamente todas as informações úteis, incluindo [tópicos do GitHub](#grooming). O CodeMeta usa [termos do Schema.org](https://schema.org/), portanto, à medida que ganha popularidade, os metadados JSON do seu pacote podem ser usados por serviços de terceiros, talvez até por mecanismos de pesquisa.
```

## Plataformas {#platforms}

> - Os pacotes devem ser executados em todas as principais plataformas (Windows, macOS, Linux).  Exceções podem ser concedidas a pacotes que interagem com funções específicas do sistema ou invólucros para utilitários que operam apenas em plataformas limitadas, mas os autores devem fazer todos os esforços para obter compatibilidade entre plataformas, incluindo compilação específica do sistema ou conteinerização de utilitários externos.

Reviewer: ```suggestion
- Os pacotes devem funcionar em todas as principais plataformas (Windows, macOS, Linux).  Pode haver exceções para pacotes que interajam com funções específicas do sistema ou que adaptem utilitários que só funcionam em plataformas limitadas, mas deve-se fazer todo o possível para garantir a compatibilidade entre plataformas, incluindo a compilação específica em cada sistema ou a containerização de utilitários externos.
```

## API do pacote {#package-api}

### Nomeação de funções e argumentos {#function-and-argument-naming}

- A nomenclatura das funções e dos argumentos deve ser escolhida de modo a trabalhar em conjunto para formar uma API de programação comum e lógica que seja fácil de ler e de autocompletar.
  
  - Considere uma `object_verb()` para as funções do seu pacote que usam um tipo de dados comum ou interagem com uma API comum. `object` refere-se aos dados/API e `verb` a ação principal.  Esse esquema ajuda a evitar conflitos de espaço de nome com pacotes que podem ter verbos semelhantes e torna o código legível e fácil de preencher automaticamente.  Por exemplo, em **stringi** as funções que começam com `stri_` manipulam strings (`stri_join()`, `stri_sort()`, e em **googlesheets** funções que começam com `gs_` são chamadas para a API do Google Sheets (`gs_auth()`, `gs_user()`, `gs_download()`).

> - Para funções que manipulam um objeto/dado e retornam um objeto/dado do mesmo tipo, faça com que o objeto/dado seja o primeiro argumento da função para aumentar a compatibilidade com os operadores de pipe (base R's `|>`, magrittr's `%>%`).

Reviewer: ```suggestion
- Para funções que manipulam um objeto/dado e retornam um objeto/dado do mesmo tipo, faça com que o objeto/dado seja o primeiro argumento da função para aumentar a compatibilidade com o operador pipe (`|>` do R base, `%>%` do pacote magrittr).
```

> - Recomendamos enfaticamente `snake_case` em vez de todos os outros estilos, a menos que esteja fazendo a portabilidade de um pacote que já esteja sendo amplamente utilizado.

Reviewer: ```suggestion
- Recomendamos fortemente usar `snake_case` em vez de todos os outros estilos, a menos que esteja fazendo a portabilidade de um pacote que já esteja sendo amplamente utilizado.
```

- Evite conflitos de nomes de funções com pacotes básicos ou outros pacotes populares (por exemplo `ggplot2`, `dplyr`, `magrittr`, `data.table`)

- A nomeação e a ordem dos argumentos devem ser consistentes entre as funções que usam entradas semelhantes.

> - As funções de pacote que importam dados não devem importar dados para o ambiente global, mas, em vez disso, devem retornar objetos. As atribuições ao ambiente global devem ser evitadas em geral.

Reviewer: ```suggestion
- As funções do pacote que importam dados não devem importar dados para o ambiente global, mas, em vez disso, devem retornar objetos. As atribuições ao ambiente global devem ser evitadas em geral.
```

### Mensagens do console {#console-messages}

> - Use o botão [pacote cli](https://cli.r-lib.org/) ou as ferramentas do R básico (`message()` e `warning()`) para se comunicar com o usuário em suas funções.

Reviewer: ```suggestion
- Use o [pacote cli](https://cli.r-lib.org/) ou as ferramentas do R básico (`message()` e `warning()`) para se comunicar com as pessoas que usam suas funções.
```

> - Os destaques do pacote cli incluem: empacotamento automático, respeito ao [convenção NO\_COLOR](https://cli.r-lib.org/articles/cli-config-user.html?q=no#no_color), muitos [elementos semânticos](https://cli.r-lib.org/articles/semantic-cli.html) e ampla documentação. Leia mais em um [postagem no blog](https://blog.r-hub.io/2023/11/30/cliff-notes-about-cli/).

Reviewer: ```suggestion
- Os destaques do pacote cli incluem: empacotamento automático, respeito a [convenção NO\_COLOR](https://cli.r-lib.org/articles/cli-config-user.html?q=no#no_color), muitos [elementos semânticos](https://cli.r-lib.org/articles/semantic-cli.html) e ampla documentação. Mais informações [neste texto](https://blog.r-hub.io/2023/11/30/cliff-notes-about-cli/).
```

> - Por favor, não use `print()` ou `cat()` a menos que seja para um `print.*()` ou `str.*()` pois esses métodos de impressão de mensagens são mais difíceis de serem suprimidos pelos usuários.

Reviewer: ```suggestion
- Por favor, não use `print()` ou `cat()` a menos que seja para um `print.*()` ou `str.*()`, pois esses métodos de impressão de mensagens são mais difíceis de serem suprimidos por quem usa a função.
```

> - Forneça uma maneira de os usuários optarem por não usar a verbosidade, de preferência no nível do pacote: torne a criação de mensagens dependente de uma variável de ambiente ou opção (como ["usethis.quiet"](https://usethis.r-lib.org/reference/ui.html?q=usethis.quiet#silencing-output) no pacote usethis), em vez de um parâmetro de função. O controle das mensagens poderia ser feito em vários níveis ("none", "inform", "debug") em vez de ser lógico (nenhuma mensagem / todas as mensagens). O controle da verbosidade é útil para usuários finais, mas também em testes. Comentários mais interessantes podem ser encontrados em um [edição do guia de design do tidyverse](https://github.com/tidyverse/design/issues/42).

Reviewer: ```suggestion
- Forneça uma maneira de suprimir a verbosidade, de preferência em nível de pacote: torne a criação de mensagens dependente de uma variável ou opção de ambiente (como ["usethis.quiet"](https://usethis.r-lib.org/reference/ui.html?q=usethis.quiet#silencing-output) no pacote usethis), em vez de um parâmetro da função. O controle das mensagens poderia ser feito em vários níveis ("nenhum", "informar", "debugar") em vez de ser lógico (nenhuma mensagem / todas as mensagens). O controle da verbosidade é útil para quem utiliza a função, mas também em testes. Você pode encontrar mais comentários interessantes [nesta issue do guia de design do tidyverse](https://github.com/tidyverse/design/issues/42)
```

> - Você pode fornecer traduções para as mensagens do seu pacote. As [potools](https://michaelchirico.github.io/potools/) R pode ajudá-lo com essa tarefa.

Reviewer: ```suggestion
- Você pode fornecer traduções para as mensagens do seu pacote. O pacote [potools](https://michaelchirico.github.io/potools/) pode te ajudar com essa tarefa.
```

### Interfaces interativas/gráficas {#interactive-graphical-interfaces}

> Se estiver fornecendo uma interface gráfica de usuário (GUI) (como um aplicativo Shiny), para facilitar o fluxo de trabalho, inclua um mecanismo para reproduzir automaticamente as etapas realizadas na GUI. Isso pode incluir a geração automática de código para reproduzir os mesmos resultados, a saída de valores intermediários produzidos na ferramenta interativa ou simplesmente um mapeamento claro e bem documentado entre as ações da GUI e as funções com script. (Consulte também ["Testes"](#testing) abaixo).

Reviewer: ```suggestion
Se estiver fornecendo uma interface gráfica de usuário (GUI) (como um aplicativo Shiny), para facilitar o fluxo de trabalho, inclua um mecanismo para reproduzir automaticamente as etapas realizadas na GUI. Isso pode incluir a geração automática de código para reproduzir os mesmos resultados, a saída de valores intermediários produzidos na ferramenta interativa ou simplesmente um mapeamento claro e bem documentado entre as ações da GUI e as funções usadas. (Consulte também ["Testes"](#testing) abaixo).
```

> O [`tabulizer` pacote](https://github.com/ropensci/tabulizer) por exemplo, tem um fluxo de trabalho interativo para extrair tabelas, mas também pode extrair apenas coordenadas para que seja possível executar novamente as coisas como um script. Além disso, dois exemplos de aplicativos brilhantes que fazem geração de código são <https://gdancik.shinyapps.io/shinyGEO/> e <https://github.com/wallaceEcoMod/wallace/>.

Reviewer: ```suggestion
O [pacote `tabulizer`](https://github.com/ropensci/tabulizer) por exemplo, tem um fluxo de trabalho interativo para extrair tabelas, mas também pode extrair apenas coordenadas para que seja possível executar novamente como um script. Além disso, dois exemplos de aplicativos brilhantes que geram código são <https://gdancik.shinyapps.io/shinyGEO/> e <https://github.com/wallaceEcoMod/wallace/>.
```

### Verificação de entrada

Recomendamos que seu pacote use um método consistente de sua escolha para [verificação de entradas](https://blog.r-hub.io/2022/03/10/input-checking/) -- seja o R básico, um pacote R ou ajudantes personalizados.

### Pacotes que envolvem recursos da Web (clientes de API)

Se o seu pacote acessar uma API da Web ou outro recurso da Web,

> - Certifique-se de que as solicitações enviem um [agente de usuário](https://httr2.r-lib.org/articles/wrapping-apis.html#user-agent) ou seja, uma maneira de identificar o que (seu pacote) ou quem enviou a solicitação. Os usuários devem ser capazes de substituir o agente de usuário padrão do pacote. Idealmente, o agente de usuário deve ser diferente nos serviços de integração contínua e no desenvolvimento (com base, por exemplo, nos nomes de usuário do GitHub dos desenvolvedores).

Author: Yes, the suggestion is to use a user agent to make an API request, so the people who develop the API can identify who made the request to the API and reach out to the package author if the package did something wrong.  It also suggested the agent should be different if the call is made when developing or the call if made by the CI.

For what I read in the Portuguese version it said the same thing that the English version. So, if you are ok, we can merge this change.

- Se o seu README tiver muitos emblemas, considere ordená-los em uma tabela HTML para facilitar aos novatos a obtenção de informações em um piscar de olhos. Veja exemplos em [`drake` repo](https://github.com/ropensci/drake) e em [`qualtRics` repo](https://github.com/ropensci/qualtRics/). As seções possíveis são
  
  - Desenvolvimento (status de CI cf [Capítulo do IC](#ci) canal do Slack para discussão, repostatus)
  - Lançamento/Publicado ([Crachás de versão do CRAN e data de lançamento do METACRAN](https://www.r-pkg.org/services#badges), [CRAN verifica o selo da API](https://github.com/r-hub/cchecksbadges), selo Zenodo)
  - Estatísticas/Uso (downloads, por exemplo [baixar emblemas do r-hub/cranlogs](https://github.com/r-hub/cranlogs.app#badges))

>     A tabela deve ser mais larga do que longa para mascarar o restante do README.

Reviewer: I'm not sure if I understand the idea of "masking the README"...
What "masking" means in this context?
@yabellini

>     A tabela deve ser mais larga do que longa para mascarar o restante do README.

Reviewer: Thank you Yani!

>     A tabela deve ser mais larga do que longa para mascarar o restante do README.

Reviewer: ```suggestion
    A tabela deve ser mais larga do que longa para não mascarar o restante do README.
```
From #959

- Se o seu pacote se conectar a uma fonte de dados ou serviço on-line, ou envolver outro software, considere que o LEIAME do seu pacote pode ser o primeiro ponto de entrada para os usuários.  Ele deve fornecer informações suficientes para que os usuários entendam a natureza dos dados, do serviço ou do software e fornecer links para outros dados e documentação relevantes.  Por exemplo,
  um LEIAME não deve se limitar a dizer: "Fornece acesso ao GooberDB", mas também incluir,
  "..., um repositório on-line de avistamentos de Goober na América do Sul.  Mais
  informações sobre o GooberDB e documentação da estrutura e dos metadados do banco de dados

>   podem ser encontrados em *link*".

Reviewer 2: Ah, entendi melhor agora, faz todo sentido! Desculpe me Bea, já retornei para a versão anterior.

- Recomendamos não criar `README.md` diretamente, mas a partir de um `README.Rmd` (um arquivo R Markdown) se você tiver algum código de demonstração. A vantagem do `.Rmd` é que você pode combinar texto com código que pode ser facilmente atualizado sempre que seu pacote for atualizado.

- Considere o uso de `usethis::use_readme_rmd()` para obter um modelo para um arquivo `README.Rmd` e para configurar automaticamente um gancho de pré-compromisso para garantir que o arquivo `README.md` seja sempre mais recente que `README.Rmd`.

- Exemplos extensos devem ser mantidos para uma vinheta. Se você quiser tornar as vinhetas mais acessíveis antes de instalar o pacote, sugerimos [criar um site para seu pacote](#website).

- Adicionar um [código de conduta e diretrizes de contribuição](#friendlyfiles).

- Veja o [`gistr` LEIAME](https://github.com/ropensci/gistr#gistr) para um bom exemplo de README a ser seguido em um pacote pequeno, e [`bowerbird` README](https://github.com/ropensci/bowerbird) para um bom exemplo de README para um pacote maior.

## Documentação {#documentation}

### Geral {#docs-general}

> - Todas as funções de pacote exportadas devem ser totalmente documentadas com exemplos.

Reviewer: ```suggestion
- Todas as funções exportadas do pacote  devem ser totalmente documentadas com exemplos.
```

If it's better to reorder, I think "exportadas d*o* pacote" would be more clear, because is related to the package that the person is developing (not any general package)

Veja [esta seção de "Escrevendo extensões R"](https://cran.rstudio.com/doc/manuals/r-release/R-exts.html#The-DESCRIPTION-file) para obter detalhes.  
Se você achar que os revisores fizeram uma contribuição substancial para o desenvolvimento do seu pacote, poderá listá-los na seção `Authors@R` com um tipo de contribuidor Revisor (`"rev"`), da seguinte forma:

```
    person("Bea", "Hernández", role = "rev",
    comment = "Bea reviewed the package (v. X.X.XX) for rOpenSci, see <https://github.com/ropensci/software-review/issues/116>"),
```

Somente inclua revisores depois de pedir seu consentimento.
Leia mais nesta postagem do blog ["Agradecendo a seus revisores: Gratidão por meio de metadados semânticos"](https://ropensci.org/blog/2018/03/16/thanking-reviewers-in-metadata/).

> Não liste os editores como colaboradores.

Reviewer 2: Não seria "não liste a equipe editorial como **colaboradora**" (no singular)?

> Não liste os editores como colaboradores.

Author: I think pessoas editoras is a better option because sometimes there are more than one editor for the same package just because how the team rotate.

> 
> As marcas registradas devem ser respeitadas.

## Licença {#licence}

O pacote precisa ter uma licença [CRAN](https://svn.r-project.org/R/trunk/share/licenses/license.db) ou [OSI](https://opensource.org/licenses) licença aceita.
O [livro de pacotes do R](https://r-pkgs.org/description.html#sec-description-authors-at-r) inclui uma útil seção [seção sobre licenças](https://r-pkgs.org/license.html).

Se o seu pacote agrupar código de outras fontes, você também precisará reconhecer os autores do código original no seu arquivo DESCIPTION, geralmente com uma função de detentor de direitos autorais: `role = "cph"`.
Para saber como atualizar seu arquivo DESCRIPTION, consulte a seção [livro de pacotes do R](https://r-pkgs.org/description.html#sec-description-authors-at-r).

## Testes {#testing}

- Todos os pacotes devem ser aprovados `R CMD check`/`devtools::check()` em todas as principais plataformas.

- Todos os pacotes devem ter um conjunto de testes que abranja a funcionalidade principal do pacote. Os testes também devem abranger o comportamento do pacote em caso de erros.

> - É uma boa prática escrever testes unitários para todas as funções e para todo o código do pacote em geral, garantindo que a funcionalidade principal seja coberta. A cobertura de testes abaixo de 75% provavelmente exigirá testes adicionais ou explicações antes de ser enviada para revisão.

Reviewer: It's more clear now!

- Se o seu pacote interagir com um banco de dados, você poderá encontrar [dittodb](https://docs.ropensci.org/dittodb) útil.

- Depois de configurar o [integração contínua (CI)](#ci) use o relatório de cobertura de código do seu pacote (cf [esta seção do nosso livro](#coverage)) para identificar linhas não testadas e adicionar mais testes.

- Mesmo que você use [integração contínua](#ci) recomendamos que você execute testes localmente antes de enviar seu pacote (talvez seja necessário definir `Sys.setenv(NOT_CRAN="true")`).

## Exemplos {#examples}

- Inclua exemplos abrangentes na documentação. Além de demonstrar como usar o pacote, eles podem funcionar como uma maneira fácil de testar a funcionalidade do pacote antes de haver testes adequados. No entanto, lembre-se de que exigimos testes em pacotes contribuídos.

- Você pode executar exemplos com `devtools::run_examples()`. Observe que quando você executa o R CMD CHECK ou equivalente (por exemplo, `devtools::check()`), seus exemplos que não estão incluídos no `\dontrun{}` ou `\donttest{}` são executados. Consulte a seção [tabela de resumo](https://roxygen2.r-lib.org/articles/rd.html#functions) nos documentos do roxygen2.

> - Para proteger os exemplos (por exemplo, exigir autenticação) a serem executados no CRAN, você precisa usar `\dontrun{}`. No entanto, para uma primeira submissão, o CRAN não permitirá que você tenha todos os exemplos escapados. Nesse caso, você pode adicionar alguns pequenos exemplos de brinquedo ou envolver o código de exemplo em `try()`. Consulte também a seção `@exampleIf` presente, no momento em que este artigo foi escrito, na versão de desenvolvimento do roxygen2.

Author: @mpadge , @maelle, can we make this change? (we should update the Spanish too)

> - Para proteger os exemplos (por exemplo, exigir autenticação) a serem executados no CRAN, você precisa usar `\dontrun{}`. No entanto, para uma primeira submissão, o CRAN não permitirá que você tenha todos os exemplos escapados. Nesse caso, você pode adicionar alguns pequenos exemplos de brinquedo ou envolver o código de exemplo em `try()`. Consulte também a seção `@exampleIf` presente, no momento em que este artigo foi escrito, na versão de desenvolvimento do roxygen2.

Reviewer 3: yes!! thank you!!

- Além de executar exemplos localmente em seu próprio computador, é altamente recomendável que você execute exemplos em um dos [sistemas de integração contínua](#ci). Mais uma vez, os exemplos que não são agrupados em `\dontrun{}` ou `\donttest{}` serão executados, mas para aqueles que são, você pode configurar suas compilações de integração contínua para executá-los por meio dos argumentos de verificação do R CMD `--run-dontrun` e/ou `--run-donttest`.

## Dependências de pacotes {#pkgdependencies}

- Em geral, é melhor ter menos dependências.

- Considere as vantagens e desvantagens envolvidas no fato de depender de um pacote como dependência. Por um lado,
  o uso de dependências reduz o esforço de codificação e pode se basear em funcionalidades úteis desenvolvidas por
  outros, especialmente se a dependência executar tarefas complexas e tiver alto desempenho,
  e/ou for bem avaliada e testada. Por outro lado, ter muitas dependências
  sobrecarrega o mantenedor para acompanhar as alterações nesses pacotes, correndo o risco de

>   para a sustentabilidade de longo prazo do seu pacote.  Isso também

Reviewer: ```suggestion
a sustentabilidade de longo prazo do seu pacote.  Isso também
```

- Verifique o status de desenvolvimento de todas as dependências que você adicionar.
  Especialmente para pacotes hospedados no GitHub, é muito útil verificar se eles são mantidos ativamente e se têm [não foram arquivados](https://ropensci.org/blog/2022/07/01/evaluating-github-activity-for-contributors/).

- Se o seu pacote (não do Bioconductor) depender de pacotes do Bioconductor, certifique-se de que as instruções de instalação no LEIAME e na vinheta sejam claras o suficiente, mesmo para um usuário que não esteja familiarizado com o ciclo de lançamento do Bioconductor.
  

>   - Se o usuário usar [`BiocManager`](https://www.bioconductor.org/install/index.html#why-biocmanagerinstall) (recomendado)? Documente isso.

Reviewer: ```suggestion
  - É necessário usar o [`BiocManager`](https://www.bioconductor.org/install/index.html#why-biocmanagerinstall) (recomendado)? Documente isso.
```

  
  - A instalação automática de pacotes do Bioconductor por `install.packages()` é suficiente? Nesse caso, mencione que o usuário precisa executar `setRepositories()` se ainda não tiver definido os repositórios do Bioconductor necessários.
  
  - Se o seu pacote depender do Bioconductor após uma determinada versão, mencione isso na DESCRIPTION e nas instruções de instalação.

- Especificar dependências mínimas (por exemplo `glue (>= 1.3.0)` em vez de apenas `glue`) deve ser uma escolha consciente. Se tiver certeza de que seu pacote quebrará abaixo de uma determinada versão de dependência, especifique-a explicitamente.
  Mas se não souber, então não há necessidade de especificar uma dependência mínima. Nesse caso, quando um usuário relatar um bug que esteja explicitamente relacionado a uma versão mais antiga de uma dependência, resolva-o.

>   Um exemplo de prática ruim seria um desenvolvedor considerar as versões de seu estado atual de dependências como a versão mínima. Isso forçaria desnecessariamente todos a atualizar (causando problemas com outros pacotes) quando não há um bom motivo por trás dessa escolha de versão.

Reviewer: ```suggestion
  Um exemplo de prática ruim seria, ao desenvolver o pacote, considerar as versões atuais de suas dependências como sendo a versão mínima. Isso forçaria desnecessariamente todos a atualizar (causando problemas com outros pacotes) quando não há um bom motivo por trás dessa escolha de versão.
```

- Na maioria dos casos em que é necessário expor as funções das dependências ao usuário, você deve importar e reexportar essas funções individuais em vez de listá-las na seção `Depends` campos.  Por exemplo, se as funções em seu pacote produzem `raster` você pode reexportar apenas as funções de impressão e plotagem do pacote **raster** do pacote raster.

- Se seu pacote usar um *sistema* do sistema, você deve
  
  - Indicá-la em DESCRIPTION;
  

>   - Verifique se ele está listado por [`sysreqsdb`](https://github.com/r-hub/sysreqsdb#sysreqs) para permitir que ferramentas automáticas o instalem, e [envie uma contribuição](https://github.com/r-hub/sysreqsdb#contributing) caso contrário;

Reviewer: ```suggestion
  - Verifique se ele está listado por [`sysreqsdb`](https://github.com/r-hub/sysreqsdb#sysreqs) para permitir que ferramentas automáticas o instalem, ou [envie uma contribuição](https://github.com/r-hub/sysreqsdb#contributing) caso contrário;
```

  
  - Verificar se há em um `configure` script ([exemplo](https://github.com/ropensci/magick/blob/c116b2b8505f491db72a139b61cd543b7a2ce873/DESCRIPTION#L19)) e fornecerá uma mensagem de erro útil se ele não puder ser encontrado ([exemplo](https://github.com/cran/webp/blob/master/configure)).
    `configure` Os scripts podem ser desafiadores, pois geralmente exigem soluções improvisadas

>     para fazer com que as diversas dependências do sistema funcionem em todos os sistemas. Use exemplos ([mais aqui](https://github.com/search?q=org%3Acran+anticonf&type=Code)) como ponto de partida, mas observe que é comum encontrar bugs e casos extremos e, muitas vezes, violar as políticas do CRAN. Não hesite em [pedir ajuda em nosso fórum](https://discuss.ropensci.org/).

Reviewer: ```suggestion
    para fazer com que as diversas dependências do sistema funcionem em todos os sistemas. Use exemplos ([mais aqui](https://github.com/search?q=org%3Acran+anticonf&type=Code)) como ponto de partida, mas observe que é comum encontrar bugs e casos extremos e, muitas vezes, violar as políticas do CRAN. Não hesite em [pedir ajuda em nosso fórum](https://discuss.ropensci.org/).
```

> ## Andaimes recomendados {#recommended-scaffolding}

Reviewer: ```suggestion
## Estruturas recomendadas {#recommended-scaffolding}
```

- Para solicitações HTTP, recomendamos o uso de [httr2](https://httr2.r-lib.org), [httr](https://httr.r-lib.org), [curl](https://jeroen.r-universe.dev/curl#) ou [crul](http://docs.ropensci.org/crul/) sobre [RCurl](https://cran.rstudio.com/web/packages/RCurl/). Se você gosta de clientes de baixo nível para HTTP, o curl é melhor, enquanto o httr2, o httr e o crul são melhores para acesso de alto nível.

- Para analisar JSON, use [jsonlite](https://github.com/jeroen/jsonlite) em vez de [rjson](https://cran.rstudio.com/web/packages/rjson/) ou [RJSONIO](https://cran.rstudio.com/web/packages/RJSONIO/).

- Para analisar, criar e manipular XML, recomendamos enfaticamente [xml2](https://cran.rstudio.com/web/packages/xml2/) para a maioria dos casos. [Você pode consultar as observações de Daniel Nüst sobre a migração de XML para xml2](https://gist.github.com/nuest/3ed3b0057713eb4f4d75d11bb62f2d66).

> - Para dados espaciais, o [sp](https://github.com/edzer/sp/) deve ser considerado obsoleto em favor do pacote [sf](https://r-spatial.github.io/sf/) e os pacotes [rgdal](https://cran.r-project.org/web/packages/rgdal/index.html), [maptools](https://cran.r-project.org/web/packages/maptools/index.html) e [rgeos](https://cran.r-project.org/web/packages/rgeos/index.html) foram aposentados em 2023. Recomendamos o uso das suítes espaciais desenvolvidas pelo [r-spatial](https://github.com/r-spatial) e [rspatial](https://github.com/rspatial) comunidades. Veja [esta edição do GitHub](https://github.com/ropensci/software-review-meta/issues/47) para discussões relevantes.

Reviewer: ```suggestion
- Para dados espaciais, o pacote [sp](https://github.com/edzer/sp/) deve ser considerado obsoleto em favor do pacote [sf](https://r-spatial.github.io/sf/) e os pacotes [rgdal](https://cran.r-project.org/web/packages/rgdal/index.html), [maptools](https://cran.r-project.org/web/packages/maptools/index.html) e [rgeos](https://cran.r-project.org/web/packages/rgeos/index.html) foram aposentados em 2023. Recomendamos o uso do conjunto de ferramentas espaciais desenvolvidas pelas comunidades [r-spatial](https://github.com/r-spatial) e [rspatial](https://github.com/rspatial). Veja [esta edição do GitHub](https://github.com/ropensci/software-review-meta/issues/47) para discussões relevantes.
```

## Controle de versão {#version-control}

- Os arquivos de origem do seu pacote devem estar sob controle de versão, mais especificamente rastreados com [Git](https://happygitwithr.com/). Você pode encontrar o [pacote gert](https://docs.ropensci.org/gert/) relevante, bem como alguns dos pacotes [usethis funcionalidade relacionada ao Git/GitHub](https://usethis.r-lib.org/reference/index.html#section-git-and-github) No entanto, você pode usar o git como quiser.

> - O nome padrão do ramo não deve ser `master` pois isso pode ser ofensivo para algumas pessoas. Consulte a seção [do projeto Git e da Software Freedom Conservancy](https://sfconservancy.org/news/2020/jun/23/gitbranchname/) para obter mais contexto. É prática geral nomear uma ramificação padrão `main` embora outros nomes também possam ser usados. Consulte a postagem do blog do tidyverse ["Renomeando a ramificação padrão"](https://www.tidyverse.org/blog/2021/10/renaming-default-branch/) para saber mais sobre como usar essa funcionalidade para ajudar a renomear as ramificações padrão.

Reviewer: ```suggestion
- O nome da ramificação (*branch*) padrão não deve ser `master`, pois isso pode ser ofensivo para algumas pessoas. Consulte a seção [do projeto Git e da Software Freedom Conservancy](https://sfconservancy.org/news/2020/jun/23/gitbranchname/) para obter mais contexto. É uma prática geral nomear uma ramificação padrão `main`, embora outros nomes também possam ser usados. Consulte a postagem do blog do tidyverse ["Renomeando a ramificação padrão"](https://www.tidyverse.org/blog/2021/10/renaming-default-branch/) para saber mais sobre como usar essa funcionalidade para ajudar a renomear as ramificações padrão.
```

- Certifique-se de listar "sucata", como `.DS_Store` em .gitignore. Você pode encontrar o arquivo [`usethis::git_vaccinate()` função](https://usethis.r-lib.org/reference/git_vaccinate.html) e a função [pacote gitignore](https://docs.ropensci.org/gitignore/) relevantes.

> - Uma seção posterior deste livro contém alguns [dicas de fluxo de trabalho do git](#gitflow).

Reviewer: ```suggestion
- Uma seção posterior deste livro contém alguns [dicas de fluxo de trabalho com git](#gitflow).
```

- Evite iniciar a descrição com o nome do pacote ou "Este pacote ...".
- Certifique-se de incluir links para sites se você envolver uma API da Web, extrair dados de um site etc. na seção `Description` do seu arquivo DESCRIPTION. Os URLs devem ser colocados entre colchetes angulares, por exemplo `<https://www.r-project.org>`.
- Em ambos os `Title` e `Description` os nomes de pacotes ou outros softwares externos devem ser colocados entre aspas simples (por exemplo, * Integração do 'Rcpp' para a biblioteca de álgebra linear com modelo 'Armadillo'*).
- Evite testes e exemplos de longa duração.  Considere `testthat::skip_on_cran` nos testes para pular coisas que demoram muito, mas ainda assim testá-las localmente e em [integração contínua](#ci).
- Inclua arquivos de nível superior, como `paper.md` arquivos de configuração de integração contínua, em seu `.Rbuildignore` arquivo.

Para obter mais dicas, consulte a lista colaborativa mantida pelo ThinkR, ["Prepare-se para o CRAN"](https://github.com/ThinkR-open/prepare-for-cran).

### Verificações do CRAN {#cranchecks}

> Quando seu pacote estiver no CRAN, ele será [verificado regularmente em diferentes plataformas](https://blog.r-hub.io/2019/04/25/r-devel-linux-x86-64-debian-clang/#cran-checks-101). As falhas nessas verificações, quando não são falsos positivos, podem levar a equipe do CRAN a entrar em contato. Você pode monitorar o estado das verificações do CRAN por meio de

Reviewer: ```suggestion
Quando seu pacote estiver no CRAN, ele será [checado regularmente em diferentes plataformas](https://blog.r-hub.io/2019/04/25/r-devel-linux-x86-64-debian-clang/#cran-checks-101). As falhas nessas verificações, quando não são falsos positivos, podem levar a equipe do CRAN a entrar em contato com você. Você pode monitorar o estado das verificações do CRAN por meio
```

> - o [`foghorn` pacote](https://fmichonneau.github.io/foghorn/).

Reviewer: ```suggestion
- do [`foghorn` pacote](https://fmichonneau.github.io/foghorn/).
```

- o [CRAN verifica os crachás](https://github.com/r-hub/cchecksbadges).

## Problemas do Bioconductor {#bioconductor-gotchas}

Se você pretende que seu pacote seja enviado para o Bioconductor ou se o pacote estiver no Bioconductor, consulte [Diretrizes de empacotamento do Bioconductor](https://www.bioconductor.org/developers/package-guidelines/) e as [livro do desenvolvedor atualizado](https://contributions.bioconductor.org/).

## Orientações adicionais {#further-guidance}

- Se você estiver enviando um pacote para o rOpenSci por meio do [repositório de revisão de software](https://github.com/ropensci/software-review) poderá encaminhar outras perguntas à equipe da rOpenSci no rastreador de problemas ou em nosso [fórum de discussão](https://discuss.ropensci.org/).

- Leia o [guia do autor](#authors-guide).

> - Leia, incorpore e aja de acordo com os conselhos do [*Guia de Colaboração* capítulo](#collaboration).

Reviewer: ```suggestion
- Leia, incorpore e aja de acordo com os conselhos do [capítulo *Guia de Colaboração*](#collaboration).
```

### Aprendendo sobre o desenvolvimento de pacotes {#learning-about-package-development}

#### Livros {#books}

- [Hadley Wickham e Jenny Bryan's *Pacotes R* livro](https://r-pkgs.org/) é um recurso excelente e de fácil leitura sobre o desenvolvimento de pacotes que está disponível para [gratuitamente on-line](https://r-pkgs.org/) (e pode ser comprado em [impresso](https://www.oreilly.com/library/view/r-packages/9781491910580/)).

> - [Escrevendo extensões do R](https://cran.r-project.org/doc/manuals/r-release/R-exts.html) é a referência canônica, geralmente a mais atualizada, para a criação de pacotes R.

Reviewer: ```suggestion
- [Escrevendo extensões do R](https://cran.r-project.org/doc/manuals/r-release/R-exts.html) é a referência canônica, geralmente a mais atualizada, para a criação de pacotes em R.
```

> - [*Dominando o desenvolvimento de software em R* por Roger D. Peng, Sean Kross e Brooke Anderson](https://bookdown.org/rdpeng/RProgDA/).

Reviewer: ```suggestion
- O livro [*Mastering Software Development in R* (Dominando o desenvolvimento de software em R) por Roger D. Peng, Sean Kross e Brooke Anderson](https://bookdown.org/rdpeng/RProgDA/) (em inglês).
```

> - [*R avançado* por Hadley Wickham](https://adv-r.hadley.nz/)

Reviewer: ```suggestion
- [*Advanced R* (R avançado) por Hadley Wickham](https://adv-r.hadley.nz/) (em inglês).
```

> - [*Guia de estilo do Tidyverse*](https://style.tidyverse.org/)

Reviewer: ```suggestion
- [*Tidyverse style guide* (Guia de estilo do Tidyverse)](https://style.tidyverse.org/)
```

- [*Guia de design do Tidyverse*](https://design.tidyverse.org/) (WIP) e o guia de design [newsletter](http://tidydesign.substack.com/).

#### Tutoriais {#tutorials}

> - [Seu primeiro pacote R em 1 hora](https://www.pipinghotdata.com/posts/2020-10-25-your-first-r-package-in-1-hour/) por Shannon Pileggi.

Reviewer: ```suggestion
- [*Your first R package in 1 hour* (Seu primeiro pacote R em 1 hora)](https://www.pipinghotdata.com/posts/2020-10-25-your-first-r-package-in-1-hour/) por Shannon Pileggi.
```

- [descrição deste fluxo de trabalho por Emil Hvitfeldt](https://www.emilhvitfeldt.com/post/2018-09-02-usethis-workflow-for-package-development/).

- [Esta foto de Matthew J Denny](https://www.mjdenny.com/R_Package_Pictorial.html).

#### Blogs {#blogs}

- [Blog do R-hub](https://blog.r-hub.io/post).

- Algumas postagens do [blog rOpenSci](https://ropensci.org/archive/) por exemplo ["Como pré-computar vinhetas de pacotes ou artigos pkgdown"](https://ropensci.org/blog/2019/12/08/precompute-vignettes/).

- Seção Package Development Corner do [boletim informativo rOpenSci](https://ropensci.org/news/).

> - Algumas postagens do [blog do tidyverse](https://www.tidyverse.org) por exemplo ["Atualizando para o testthat edition 3"](https://www.tidyverse.org/blog/2022/02/upkeep-testthat-3/).

Reviewer: ```suggestion
- Algumas postagens do [blog do tidyverse](https://www.tidyverse.org), por exemplo [*Upgrading to testthat edition 3* ("Atualizando para o testthat edition 3")](https://www.tidyverse.org/blog/2022/02/upkeep-testthat-3/).
```
